CSG "inner" timeseries function API and models for it

mikapfl opened the pull request:
# Pull request

Please confirm that this pull request has done the following:

- [ ] Tests added
- [ ] Documentation added (where applicable)
- [ ] Description in a ``.rst`` file in the directory ``changelog_unreleased`` added - remember to include a ``*`` to make it a bullet point

## Description

WIP

mikapfl commented:
@JGuetschow 
Regarding our idea to pre-allocate the result arrays, I did some micro-benchmarks to see if memory allocation is a relevant factor.

**Abstract**:
Allocating memory is unavoidable even for simple algorithms, and using xarray's built-in (not-in-place) functions where possible is faster than other alternatives. I think we should use non-in-place (i.e. copy-on-write) semantics everywhere.

**Results**:
In my test, I used a timeseries spanning from 1850 to 2022. It had one NaN entry, and I copied it (before the benchmarking) 40 000 times. I filled the NaN with another timeseries also spanning from 1850 to 2022. Algorithms I benchmarked:

```python
# not in place
def nip():
    res = []
    for ts in tss:
        res.append(ts.fillna(fts))
    return res

# not in place, using join="exact"
# this avoids potentially costly alignment computations when ts and fts already have the same shape
def nip_je():
    res = []
    for ts in tss:
        res.append(xr.core.ops.fillna(ts, fts, join="exact"))
    return res

# in-place
# note we have to allocate memory in the shape of `ts` for `tsn`. I couldn't find a formulation
# to do this without allocating memory and without looping
def ip():
    for ts in tss:
        tsn = ts.isnull()
        ts[tsn] = fts[tsn]

# in-place cheating (we already know where the NaN is)
# just a baseline to know what an absolutely minimal in-place loop would do
def ip_cheat():
    for ts in tss:
        ts[0] = fts[0]
```

Results for `len(tss) = 40000`:

| algorithm | execution time |
| -------------- | ---------------------- |
| `nip` | 8.92 s |
| `nip_je` | 5.12 s |
| `ip` | 18.3 s |
| `ip_cheat` | 2.94 s |

As you can see, the fastest not-in-place algorithm `nip_je` takes less than twice the time that the cheating "let's just skip actually checking for NaNs" `ip_cheat` algorithm takes. The `ip` algorithm which updates the `ts` in-place is even a lot slower than the `fillna` formulation. And generally, everything is pretty fast.

**Conclusion**:

Let's not complicate things to avoid allocating memory. Especially early on, simplicity trumps performance every time, and the performance wins are either small or not there at all. The default mode of xarray and the only mode of the rest of primap2 is copy-on-write, so let's do that. If we find out later that the performance is not good enough, we can measure with actual data and then optimize what needs optimization. If it then turns out we need to avoid allocating memory, we can change our internal APIs and algorithms relatively easily.

My intuition was (like yours) that avoiding to allocate memory should be quite nice. However, I think I got this idea from when I worked with numba, i.e. compiled code. So, when you already squeezed the 95 % performance wins you can get from all kinds of other optimizations. In a small test I did doing this via numpy straight instead of xarray, it can be done in 44 milliseconds, so orders of magnitude faster - but without metadata etc. So if we find performance is not good enough and the filling strategies are to blame, I'd probably look into something which uses raw numpy (but then you again pay the price of not having metadata handled automatically).

mikapfl commented:
The via-numpy algorithm which is orders of magnitude faster (44 ms):
```python
# in-place via directly accessing the underlying numpy arrays
# avoids the xarray overhead, but also all the xarrays alignment checks
# and duck typing etc (e.g., won't work with dask arrays)
def ip_via_numpy(tss, fts):
    for ts in tss:
        tsn = np.isnan(ts.values)
        ts.values[tsn] = fts.values[tsn]
```

JGuetschow commented:
OK, that's good. I had merge in mind when thinking of the performance issues but as we know that all time series are for different fixed coordinate tuples append works. I think it doesn't matter if it's 44ms or 18s if that's all the time the main loop needs for data collection.

JGuetschow commented:
Nice that's it#s making fast progress. Some quick comments.

>     * I've now added an array which contains a string representation of the source of every single datapoint to the API.
I think it#s sufficient (actually better) to have the information once per result time series. Something like "Years 1990-2021 from CRF2023, 1965-1989 and 2022 from EDGAR 8.0 matched with XXX, scaled using a=X, b=Y, ...". The per data point info is harder to read for humans and with an assumed 200 bytes per data point it's 2GB of metadata which is much more than the actual data. It's also easier to specify the scaling calculations as they are global anyway.

> Todos:
> 
>     * Think a bit more about the traceability use case, probably I'll explore the possibility space a bit. Current ideas: structured logging, `uncertaintes`.
See my comment above, maybe that makes it easier.

>     * strategies should go into a submodule or something
agree

>     * For deciding how we deal with the traceability use case, it would be great to have a test case with realistic data proportions. This would mean adding the main `compose` function (at least in a rudimentary form) and something which generates input_data etc. Maybe this should actually be the next step.

I'll generate a test dataset when I'm back at work, either the week after next week, or end of next week.

mikapfl commented:
> I think it#s sufficient (actually better) to have the information once per result time series.

Yeah, my initial intuition was that having a "metadata array" would be easier to produce and turning it into something human-readable would be a second step. After implementing it for the trivial filling strategy, I also suspect another solution is better (like producing human-readable stuff on a timeseries level directly). But I'll need some more exploration to fully convince myself.

mikapfl commented:
I added a first prototype (enough to start thinking about it) of the "outer" `compose` algorithm which iterates over all non-priority coordinates.

It doesn't work yet, I have to add proper DataArray merging (I hoped `xr.combine_by_coords` would be smart enough, but with the `opulent_ds` example ds, it was too confused by the 10 or so dimensions). As discussed, a simple `concat` should do it, it just has to be at the right places.

I chose to hand-roll a recursive algorithm instead of using xarray's `groupby` functionality - maybe that's not correct, but good enough at this stage.

mikapfl commented:
Okay, the first prototype of the `compose` function now works, and I added a "real-size" performance test for the current implementation which deals with lots of strings. It is too slow. Just computing CO2 (but all category and countries, with 10 sources) with straight substitution only didn't finish in 15 minutes. We could probably make it faster, but I'll instead explore the other options first. As you already identified, the memory usage is also high for the "array of strings" architecture, and it is not nice to work with.

mikapfl commented:
After measuring the performance of the "array of strings" strategy, I backtracked a bit and started measuring a version without any traceability first. I.e., a prototype which just calculates a dataset in the shape of primap-hist, without source info.

What I learned so far:
* `xr.concat` is actually too slow. It does all kinds of sanity checks which are not necessary in our case because we know the shape of our result beforehand. So, I added pre-allocation of the result in the **outer** loop.
* I now have a test case which has a somewhat involved formulation of priorities (200 country-specifc exceptions), but all sources have no NaNs, so in effect it just copies the highest-priority source (per timeseries) to the result. With pre-allocation of the result, it takes 3.5 minutes to calculate CO2 only.

So, it looks like we have to focus somewhat on performance. Meeting our performance goals will not be possible without careful performance work. I'm a little surprised about that, but, well, the input data (with 10 sources in the shape of primap-hist) is 1 GB of data. That's a lot of numbers to crunch, even if you just copy bits from different places into a new array.

My plan is now to do some performance work on the non-tracing version, then add tracing and continue measuring. This way, we can actually know how much overhead tracing adds.

mikapfl commented:
Haha, this is a case again of "measure performance with care":

Profiling shows that about half of the time is spent in `__getitem__`, and one third in `__setitem__`, which sounds about right for something which mostly searches for data in one array, then copies it into another array. However, going a bit deeper in the callstack, it shows that both `__getitem__` and `__setitem__` call `ndim` to check the number of dimensions - and 90 % of the total time is spent in `ndim`. Curios. Well, it turns out that our arrays have units, and `pint` has an implementation of `ndim` which stringifys the whole array. Fixing this, computing CO2 only takes 14 seconds instead of 3.5 minutes.

mikapfl commented:
Hi,

I now arrived at a prototype implementation (over in the [structured logging branch](https://github.com/pik-primap/primap2/tree/csg-strctured-logging)) which I like quite a lot. It is essentially a mix between "array of strings" and "structured logging / one giant string".
The idea is: We have one processing description (which boils down to a string) *per timeseries*. This means we don't have to repeat so much information, we can essentially just say "for years 1850-1870: did x". This information is then held in the result dataset as distinct variables. It looks like this:

```python
In [1]: import primap2

In [2]: ds = primap2.open_dataset("perf_test_result.nc")

In [3]: ds
Out[3]: 
<xarray.Dataset> Size: 49MB
Dimensions:                     (time: 273, category (IPCC2006_PRIMAP): 24,
                                 area (ISO3): 215, scenario (PRIMAP-hist): 2,
                                 provenance: 1)
Coordinates:
  * time                        (time) datetime64[ns] 2kB 1750-01-01 ... 2022...
  * category (IPCC2006_PRIMAP)  (category (IPCC2006_PRIMAP)) <U8 768B '0' ......
  * area (ISO3)                 (area (ISO3)) <U9 8kB 'ABW' 'AFG' ... 'ZWE'
  * scenario (PRIMAP-hist)      (scenario (PRIMAP-hist)) <U6 48B 'HISTCR' 'HI...
  * provenance                  (provenance) <U8 32B 'measured'
Data variables:
    CO2                         (time, category (IPCC2006_PRIMAP), area (ISO3), scenario (PRIMAP-hist), provenance) float64 23MB ...
    Processing of CO2           (category (IPCC2006_PRIMAP), area (ISO3), scenario (PRIMAP-hist), provenance) <U644 27MB ...

In [4]: ds["Processing of CO2"]
Out[4]: 
<xarray.DataArray 'Processing of CO2' (category (IPCC2006_PRIMAP): 24,
                                       area (ISO3): 215,
                                       scenario (PRIMAP-hist): 2, provenance: 1)> Size: 27MB
array([[[["In time=all: used values from 'source #0'"],
         ["In time=all: used values from 'source #0'"]],

        [["In time=all: used values from 'source #0'"],
         ["In time=all: used values from 'source #0'"]],

        [["In time=all: used values from 'source #0'"],
         ["In time=all: used values from 'source #0'"]],

        ...,

        [["In time=all: used values from 'source #0'"],
         ["In time=all: used values from 'source #0'"]],

        [["In time=all: used values from 'source #0'"],
         ["In time=all: used values from 'source #0'"]],

        [["In time=all: used values from 'source #0'"],
         ["In time=all: used values from 'source #0'"]]],

...

       [[["In time=all: used values from 'source #0'"],
         ["In time=all: used values from 'source #0'"]],

        [["In time=all: used values from 'source #0'"],
         ["In time=all: used values from 'source #0'"]],

        [["In time=all: used values from 'source #0'"],
         ["In time=all: used values from 'source #0'"]],

        ...,

        [["In time=all: used values from 'source #0'"],
         ["In time=all: used values from 'source #0'"]],

        [["In time=all: used values from 'source #0'"],
         ["In time=all: used values from 'source #0'"]],

        [["In time=all: used values from 'source #0'"],
         ["In time=all: used values from 'source #0'"]]]], dtype='<U644')
Coordinates:
  * category (IPCC2006_PRIMAP)  (category (IPCC2006_PRIMAP)) <U8 768B '0' ......
  * area (ISO3)                 (area (ISO3)) <U9 8kB 'ABW' 'AFG' ... 'ZWE'
  * scenario (PRIMAP-hist)      (scenario (PRIMAP-hist)) <U6 48B 'HISTCR' 'HI...
  * provenance                  (provenance) <U8 32B 'measured'

In [5]: ds["Processing of CO2"].pr.loc[{"category": "0", "area": "ABW"}]
Out[5]: 
<xarray.DataArray 'Processing of CO2' (scenario (PRIMAP-hist): 2, provenance: 1)> Size: 5kB
array([["In time=all: used values from 'source #0'"],
       ["In time=all: used values from 'source #0'"]], dtype='<U644')
Coordinates:
    category (IPCC2006_PRIMAP)  <U8 32B '0'
    area (ISO3)                 <U9 36B 'ABW'
  * scenario (PRIMAP-hist)      (scenario (PRIMAP-hist)) <U6 48B 'HISTCR' 'HI...
  * provenance                  (provenance) <U8 32B 'measured'
```

I think the advantage versus a giant string is that it is relatively easy to process this further, e.g. to add the processing description to the change HTML you produce, Johannes. In a giant string, we have to parse the string to then search for the information on the country, category, and scenario we are interested in. In contrast, adding this info as timeseries-level strings into the DataArray, we can use the same techniques we already have for selecting data to select the processing info:

```python
In [6]: ds.pr.loc[{"category": "0", "area": "ABW"}]
Out[6]: 
<xarray.Dataset> Size: 12kB
Dimensions:                     (time: 273, scenario (PRIMAP-hist): 2,
                                 provenance: 1)
Coordinates:
  * time                        (time) datetime64[ns] 2kB 1750-01-01 ... 2022...
    category (IPCC2006_PRIMAP)  <U8 32B '0'
    area (ISO3)                 <U9 36B 'ABW'
  * scenario (PRIMAP-hist)      (scenario (PRIMAP-hist)) <U6 48B 'HISTCR' 'HI...
  * provenance                  (provenance) <U8 32B 'measured'
Data variables:
    CO2                         (time, scenario (PRIMAP-hist), provenance) float64 4kB ...
    Processing of CO2           (scenario (PRIMAP-hist), provenance) <U644 5kB ...
```

Next, I'll look into the uncertainties package.

mikapfl commented:
A prototype implementation using [uncertainties](https://pythonhosted.org/uncertainties/numpy_guide.html) is now in the [csg-uncertainties branch](https://github.com/pik-primap/primap2/tree/csg-uncertainties). Things I learned about the uncertainties implementation:
* Given the additional functionality (full Gaussian error propagation), the implementation was surprisingly straight-forward. You can put just stick uncertainties' uarrays into xarray and things work.
* However, not all xarray functions are handled correctly transparently, e.g. xarray's `fillna` can't deal with ufloats. You have to drop down to (u)numpy functions and do it a bit more manually. This is something that every filling strategy will have to deal with.
* Performance is actually pretty good! Full input data in the shape of primap-hist is computed *for all entities* in 1.5 min. Compare that to the "structured logging" approach, which takes 7.5 min for the same task (both without serious optimization work - but optimization means uglifying the filling strategy implementations, so has to be done with care). However, the comparison is not quite fair because the "structured logging" approach hands you back consumption-ready strings, while the uncertainties approach hand you back ufloats and you have to still create consumption-ready strings. So, its not apples to apples. The main point is: The uncertainties approach is not slow at all. However, this concerns substitution only, where no actual computations with ufloats have to be done - computations on ufloats are of course slower than computations on floats.
* You can't save full ufloats to netcdf. You can save the nominal value and standard deviation to netcdf, but if you want the correlations and everything (including the source information), you have to pickle. That's unfortunate.
* The memory consumption is much higher for the uncertainties approach. A full run in the shape of primap-hist peaks at about 10 GiB memory, while a "structured logging" run in the same shape peaks at about 3 GiB. Additionally, saving to netcdf takes essentially no additional memory, while pickling the uncertainties result after computing it peaks at a total of 18 GiB memory consumption.

I think we are in a good position to decide tomorrow.

mikapfl commented:
We decided not to use the `uncertainties` package for now, strings are more practical as a description for processing steps on a timeseries-level for now. I updated the `csg` branch (and thus this PR) to this implementation.

mikapfl commented:
@JGuetschow I propose that we merge this soon (after I added some documentation). Additional tests, filling strategies, saving / string conversion functions, and configuration options can be added in follow-up PRs. I think it is easier to do this in smaller PRs and this thing here is already a solid chunk of code and functionality.

After I added a few more detailed tests, I feel pretty confident that there are no major bugs left.

mikapfl commented:
@JGuetschow From my side, this PR is finished now. I'd add other features (like prioritising and selecting along the entity dimension) in another PR.

JGuetschow commented on primap2/tests/test_units.py:
```diff
@@ -57,6 +57,16 @@ def test_convert_to_mass(opulent_ds: xr.Dataset):
     assert_equal(da_converted, da_expected)
 
 
+def test_convert_rount_trip(opulent_ds: xr.Dataset):
```
typo in function name? (roun*t*_trip)

JGuetschow reviewed (approved):
Looks great. I made some minor comments, but nothing critical.

mikapfl commented on primap2/csg/_models.py:
```diff
@@ -0,0 +1,168 @@
+        dimensions (fixed dimensions) to limit the selection to specific cases.
+        Examples:
+        [{"area (ISO3)": "COL", "source": "A"}, {"source": "B"}]
+        would select source "A" for Columbia, but source "B" for all other countries.
```
Yes, that's correct, it will select source "A" as first priority for Columbia, and source "B" as second priority, but it would select only source "B" for all other countries. I'll update the description.

mikapfl commented on primap2/csg/_models.py:
```diff
@@ -0,0 +1,168 @@
+        Example: [({"source": ["FAOSTAT", "UNFCCC]}, StraightStrategy()),
+                  ({}, GlobalStrategy())]
+        Note that the strategy can depend on fixed coordinates as well as priority
+        coordinates.
```
Yep, good idea, will do.

JGuetschow commented on primap2/csg/_models.py:
Your reasoning makes sense. But I'm supporting renaming as the "selection" might confuse not only me.
